packages/rest-crud/src/crud-rest.controller.ts

```diff
@@ -0,0 +1,320 @@
+// Node module: @loopback/crud-rest
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {
+  Count,
+  CountSchema,
+  DataObject,
+  Entity,
+  EntityCrudRepository,
+  Filter,
+  Where,
+} from '@loopback/repository';
+import {
+  api,
+  del,
+  get,
+  getFilterSchemaFor,
+  getJsonSchema,
+  getModelSchemaRef,
+  getWhereSchemaFor,
+  JsonSchemaOptions,
+  jsonToSchemaObject,
+  MediaTypeObject,
+  param,
+  ParameterObject,
+  patch,
+  post,
+  requestBody,
+  ResponsesObject,
+  SchemaObject,
+} from '@loopback/rest';
+
+// Ideally, this file should simply `export class CrudRestController<...>{}`
+// Unfortunately, that's not possible for several reasons.
+//
+// First of all, to correctly decorate methods and define schemas for request
+// and response bodies, we need to know the target model which will be used by
+// the controller. As a result, this file has to export a function that will
+// create a constructor class specific to the given model.
+//
+// Secondly, TypeScript does not allow decorators to be used in class
+// expressions - see https://github.com/microsoft/TypeScript/issues/7342.
+// As a result, we cannot write implement the factory as `return class ...`,
+// but have to define the class as an internal type and return the controller
+// constructor in a new statement.
+// Because the controller class is an internal type scoped to the body of the
+// factory function, we cannot use it to describe the return type. We must
+// explicitly provide the return type.
+//
+// To work around those issues, we use the following design:
+// - The interface `CrudRestController` describes controller methods (members)
+// - The type `CrudRestControllerCtor` describes the class constructor.
+// - `defineCrudRestController` returns `CrudRestControllerCtor` type.
+
+/**
+ * This interface describes prototype members of the controller class
+ * returned by `defineCrudRestController`.
+ */
+export interface CrudRestController<
+  T extends Entity,
+  IdType,
+  IdName extends keyof T,
+  Relations extends object = {}
+> {
+  /**
+   * The backing repository used to access & modify model data.
+   */
+  readonly repository: EntityCrudRepository<T, IdType>;
+
+  /**
+   * Implementation of the endpoint `POST /`.
+   * @param data Model data
+   */
+  create(data: Omit<T, IdName>): Promise<T>;
+}
+
+/**
+ * Constructor of the controller class returned by `defineCrudRestController`.
+ */
+export interface CrudRestControllerCtor<
+  T extends Entity,
+  IdType,
+  IdName extends keyof T,
+  Relations extends object = {}
+> {
+  new (
+    repository: EntityCrudRepository<T, IdType, Relations>,
+  ): CrudRestController<T, IdType, IdName, Relations>;
+}
+
+/**
+ * Options to configure different aspects of a CRUD REST Controller.
+ */
+export interface CrudRestControllerOptions {
+  /**
+   * The base path where to "mount" the controller.
+   */
+  basePath: string;
```

Comment:
Is it possible to make this optional and use a default `basePath` if one isn't provided?

Reply:
That would require us to implement some logic to convert a model name (e.g. `Product`) to a path (e.g. `/products`). This gets quickly tricky - do we want `/people` or `/persons` for `Person` model?

I'd prefer to keep `basePath` required for this initial implementation, we can make it optional later.

Reply:
Perhaps the purpose of this field is not clear enough and we should improve the api docs? Can you suggest a better wording? Would an example help?

Reply:
> That would require us to implement some logic to convert a model name (e.g. `Product`) to a path (e.g. `/products`). This gets quickly tricky - do we want `/people` or `/persons` for `Person` model?

I was thinking we could use something similar to [how the controller generator does it](https://github.com/strongloop/loopback-next/blob/feat/rest-crud/packages/cli/generators/controller/index.js#L191) which in this scenario would make it `/people`.

> I'd prefer to keep `basePath` required for this initial implementation, we can make it optional later.

That's fine with me 👍 

> Perhaps the purpose of this field is not clear enough and we should improve the api docs? Can you suggest a better wording? Would an example help?

It's clear to me. My suggestion was so that we could make [`options`](https://github.com/strongloop/loopback-next/pull/3582/files#diff-fac7093db3933c6e14d7bd88c107299eR134) optional and be able to exclude it from the function call:
```ts
   const CrudRestController = defineCrudRestController<
     Product,
     typeof Product.prototype.id,
     'id'
   >(Product, {basePath: '/products'});
   ```
vs.
```ts
   const CrudRestController = defineCrudRestController<
     Product,
     typeof Product.prototype.id,
     'id'
   >(Product);
   ```

Reply:
> I was thinking we could use something similar to [how the controller generator does it](https://github.com/strongloop/loopback-next/blob/feat/rest-crud/packages/cli/generators/controller/index.js#L191) which in this scenario would make it `/people`.

Makes sense! I had mixed experience with auto-generated base path for CRUD endpoints in LB3. Some people found the pluralization rules confusing ("people" vs. "persons"), some wanted different mechanism for converting pascal-cased model names (e.g. "ProductReview") to HTTP paths (e.g. "/ProductReviews", "/product-reviews", etc.).

Let's keep `basePath` optional for now and wait until we learn more about this area while working on https://github.com/strongloop/loopback-next/issues/2036.